### PR TITLE
Add spec for team endpoints

### PR DESCRIPTION
**Pull request by se7entyse7en**
Signed-off-by: Lou Marvin Caraig <loumarvincaraig@gmail.com>

### CODE REVIEW

**Comment by se7entyse7en** [comment]
I'll continue development starting from this and eventually fix according to issues you might raise.

File: server/athenian/api/openapi/openapi.yaml [hunk 1]
```diff
@@ -838,6 +838,183 @@ paths:
...
+      x-openapi-router-controller: athenian.api.controllers.team_controller
+      x-codegen-request-body-name: body
+    put:
+      operationId: update_team
+      parameters:
+      - description: Numeric identifier of the team to update, empty on create.
+        explode: false
+        in: path
+        name: id
+        # Not actually required, but cannot set optional path parameter:
```
Comment (se7entyse7en) [inline server/athenian/api/openapi/openapi.yaml:930]: @vmarkovtsev wdyt about this? Do you have a better approach?

**Comment by warenlg** [comment]
Wow, I'm curious to see how we label this PR...

**Comment by warenlg** [comment]
Review Required, not bad!

File: server/athenian/api/openapi/openapi.yaml (hunk 1 above)
Comment (vmarkovtsev) [reply server/athenian/api/openapi/openapi.yaml:930]: You set it as `required: true` so we will not be able to call `PUT /team` to create a new one. 

I had a similar problem with `/become` and I solved it with a URL parameter. However, we've got a CRUD here...

So maybe I added `CREATE /reposet` because of this issue.

File: server/athenian/api/openapi/openapi.yaml [hunk 2]
```diff
@@ -838,6 +838,183 @@ paths:
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The specified team was not found.
+        409:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: There is an existing team with the same contents.
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:966]: Yes, this is a very good check to avoid duplicate teams in our precious DB :+1:

File: server/athenian/api/openapi/openapi.yaml [hunk 3]
```diff
@@ -1083,6 +1260,90 @@ components:
       items:
         $ref: '#/components/schemas/DeveloperID'
       type: array
+    TeamMember:
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1263]: I wonder if we can merge `TeamMember` with `User` which are reported in `/account`...

File: server/athenian/api/openapi/openapi.yaml [hunk 4]
```diff
@@ -1083,6 +1260,91 @@ components:
...
+      description: A team member.
+      example:
+        login: github.com/gkwillie
+        id: github|60340680
+        native_id: 60340680
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1268]: Let's not add `native_id` to the response, I cannot imagine why it would needed, and you'd better not know [how we get it](https://github.com/athenianco/athenian-api/blob/master/server/athenian/api/auth.py#L350) :smile:

File: server/athenian/api/openapi/openapi.yaml [hunk 5]
```diff
@@ -1083,6 +1260,91 @@ components:
...
+        name: Groundskeeper Willie
+        email: bot@athenian.co
+        picture: https://avatars0.githubusercontent.com/u/60340680?v=4
+        updated: 2020-01-23T12:00:00Z
+      allOf:
+        - $ref: '#/components/schemas/User'
+        - properties:
+            login:
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1276]: How about adding this to the `User` model? Thus we can delete this `allOf`.

File: server/athenian/api/openapi/openapi.yaml [hunk 6]
```diff
@@ -1083,6 +1260,91 @@ components:
...
+              $ref: '#/components/schemas/DeveloperID'
+          required:
+            - login
+    Team:
+      description: A list of members part of the same team.
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1281]:
```suggestion
      description: List of team members.
```

File: server/athenian/api/openapi/openapi.yaml [hunk 7]
```diff
@@ -1083,6 +1260,91 @@ components:
...
+          updated: 2020-01-23T12:00:00Z
+        - login: github.com/se7entyse7en
+          id: github|5599208
+          native_id: 5599208
+          name: Lou Marvin Caraig
+          email: loumarvincaraig@gmail.com
+          picture: https://avatars1.githubusercontent.com/u/5599208?v=4
+          updated: 2020-05-05T18:19:05Z
+      properties:
+        id:
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1301]:
```suggestion
        id:
          description: Team identifier.
```

File: server/athenian/api/openapi/openapi.yaml [hunk 8]
```diff
@@ -1083,6 +1260,91 @@ components:
...
+          type: integer
+        name:
+          type: string
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1304]:
```suggestion
          description: Name of the team.
          type: string
```

File: server/athenian/api/openapi/openapi.yaml [hunk 9]
```diff
@@ -838,6 +838,183 @@ paths:
...
+        schema:
+          type: integer
+        style: simple
+      responses:
+        200:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Team'
+          description: List of members in the specified team.
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:899]:
```suggestion
          description: List of the members in the specified team.
```

File: server/athenian/api/openapi/openapi.yaml [hunk 10]
```diff
@@ -838,6 +838,183 @@ paths:
...
+          description: The user is forbidden to list the specified team.
+        404:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The specified team was not found.
+      security:
+      - bearerAuth: []
+      summary: List a team's members. The user must be in the account that owns the
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:914]:
```suggestion
      summary: List the team's members. The user must belong to the account that owns the
```

File: server/athenian/api/openapi/openapi.yaml [hunk 11]
```diff
@@ -838,6 +838,183 @@ paths:
...
+      - team
+# {% if False %}
+      - default
+# {% endif %}
+      x-openapi-router-controller: athenian.api.controllers.team_controller
+      x-codegen-request-body-name: body
+    put:
+      operationId: update_team
+      parameters:
+      - description: Numeric identifier of the team to update, empty on create.
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:926]:
```suggestion
      - description: Numeric identifier of the team to update, empty on creation.
```

File: server/athenian/api/openapi/openapi.yaml (hunk 1 above)
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:930]: Got it, it makes sense. I'll then add an endpoint similarly to `/reposet`

File: server/athenian/api/openapi/openapi.yaml (hunk 5 above)
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:1276]: Indeed, probably that would the best thing. 👍

File: server/athenian/api/openapi/openapi.yaml (hunk 4 above)
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:1268]: hahaha 😂 

Anyway, the only reason I put is to reuse `User` and it's a required property in `User`,

Comment (vmarkovtsev) [reply server/athenian/api/openapi/openapi.yaml:1339]: ok, then let's not touch it.

**Comment by se7entyse7en** [comment]
@vmarkovtsev all issues should have been addressed, and I added the endpoint for team creation

**Comment by se7entyse7en** [comment]
But I guess this cannot be merged until I do not generate the models and add the new `login` attribute in `User`.

File: server/athenian/api/models/web/team.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+from typing import List, Dict, Type
+
+from athenian.api.models.base_model_ import Model
+from athenian.api.models.user import User
+from athenian.api import util
+
+
+class Team(Model):
+    """NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team.py:13]: Can you please copy-paste the model description here

File: server/athenian/api/models/web/team.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
+# coding: utf-8
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team.py:1]:
```suggestion
```
(we don't need it in py3)

File: server/athenian/api/models/web/team.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+from athenian.api import util
+
+
+class Team(Model):
+    """NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
+
+    Do not edit the class manually.
+    """
+
+    def __init__(self, name: str = None, members: List[User] = None):
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team.py:18]:
```suggestion
    def __init__(self, name: Optional[str] = None, members: Optional[List[User]] = None):
```

File: server/athenian/api/models/web/team.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+    @classmethod
+    def from_dict(cls, dikt: dict) -> "Team":
+        """Returns the dict as a model
+
+        :param dikt: A dict.
+        :return: The Team of this Team.
+        """
+        return util.deserialize_model(dikt, cls)
+
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team.py:39]:
```suggestion
```

File: server/athenian/api/models/web/team.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    @property
+    def name(self):
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team.py:41]:
```suggestion
    def name(self) -> str:
```

File: server/athenian/api/models/web/team.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+        """Gets the name of this Team.
+
+        Name of the team.
+
+        :return: The name of this Team.
+        :rtype: str
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team.py:47]:
```suggestion
```

File: server/athenian/api/models/web/team.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+        """
+        return self._name
+
+    @name.setter
+    def name(self, name):
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team.py:52]:
```suggestion
    def name(self, name: str):
```
... and so on. The default generated code is not ideal.

File: server/athenian/api/models/web/user.py [hunk 19]
```diff
@@ -131,7 +136,30 @@ def id(self, id: str):
         self._id = id
 
     @property
-    def native_id(self) -> str:
+    def login(self):
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/user.py:139]:
```suggestion
    def login(self) -> str:
```

File: server/athenian/api/openapi/openapi.yaml [hunk 20]
```diff
@@ -1083,6 +1296,139 @@ components:
...
+        team:
+          $ref: '#/components/schemas/Team'
+      required:
+      - account
+      - team
+      type: object
+    TeamID:
+      description: Team identifier.
+      example: Engineering
+      type: string
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1333]:
```suggestion
```
(not used)

File: server/athenian/api/openapi/openapi.yaml [hunk 21]
```diff
@@ -838,6 +838,219 @@ paths:
...
+  /team/create:
+    post:
+      operationId: create_team
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/TeamCreateRequest'
+        x-body-name: body
+      responses:
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply server/athenian/api/openapi/openapi.yaml:850]: How about 422 Unprocessable Entity. We raise 400 on JSON syntax errors, so if we can find a more specific code for this error then it will be distinguishable.

File: server/athenian/api/openapi/openapi.yaml [hunk 22]
```diff
@@ -838,6 +838,219 @@ paths:
...
+      summary: Create a team.
+      tags:
+      - team
+# {% if False %}
+      - default
+# {% endif %}
+      x-openapi-router-controller: athenian.api.controllers.team_controller
+      x-codegen-request-body-name: body
+  /team/{id}:
+    delete:
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply server/athenian/api/openapi/openapi.yaml:880]: We are likely to use ARRAY the same way as in reposets to migrate less tables, though IDK exactly how Marvin sees it.

File: server/athenian/api/openapi/openapi.yaml (hunk 21 above)
[the remark this answers is not in the document]
Comment (dpordomingo) [reply server/athenian/api/openapi/openapi.yaml:850]: `422` lgtm

File: server/athenian/api/openapi/openapi.yaml [hunk 23]
```diff
@@ -838,6 +838,219 @@ paths:
...
+          application/json:
+            schema:
+              $ref: '#/components/schemas/TeamCreateRequest'
+        x-body-name: body
+      responses:
+        200:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/CreatedIdentifier'
```
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:855]: yup totally makes sense to me, I just almost blindly copied what was being sent during the `PUT`, gonna change it since all those info are available to the api 👍

File: server/athenian/api/openapi/openapi.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (dpordomingo) [reply server/athenian/api/openapi/openapi.yaml:880]: A problem I see if we embed the users in the teams is that they can be outdated easily if we do not implement other update processes.
Also, if at some point we need to change their schema, the migration process would be also more complex.

File: server/athenian/api/openapi/openapi.yaml (hunk 21 above)
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:850]: good point, I'd just go for 400 with different error message. I think that distinguishing these two different cases by different status code will be important when and if we're going to have clients different from the webapp.

File: server/athenian/api/openapi/openapi.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:880]: Yup, I'd rather user an ARRAY as we do for `reposets` 👍

File: server/athenian/api/openapi/openapi.yaml [hunk 24]
```diff
@@ -838,6 +838,219 @@ paths:
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The user does not belong to the specified account.
+        409:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: Concurrent requests that are finishing the account registration.
```
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:1043]: mhmmm tricky. tbh I added this to make as much similar to `/reposets/{id}`, but probably in this case even `200` with empty list would fit as well. I think that `409` in case of non-finished account registration fits for enpdoints involving write ops.

File: server/athenian/api/openapi/openapi.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (dpordomingo) [reply server/athenian/api/openapi/openapi.yaml:880]: are we embedding the whole user or only its id?

File: server/athenian/api/openapi/openapi.yaml (hunk 24 above)
[the remark this answers is not in the document]
Comment (dpordomingo) [reply server/athenian/api/openapi/openapi.yaml:1043]: Agree on returning empty list for read ops, and error on write ones.

File: server/athenian/api/openapi/openapi.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply server/athenian/api/openapi/openapi.yaml:880]: Only the ids - or, more precisely, their logins, if we allow out-of-account members

[the remark this answers is not in the document]
Comment (dpordomingo) [reply server/athenian/api/openapi/openapi.yaml:880]: then lgtm.
(`login` can change, so it might be not the best way to identify the users in the embedded list)

[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:880]: I still need to better see how this could be done, do you need the answer now? I don't think that the way it is going to implemented will change the spec.

File: server/athenian/api/openapi/openapi.yaml (hunk 23 above)
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:855]: done [here](https://github.com/athenianco/athenian-api/pull/334/commits/b197c83d65ad261309a3cdb0c546443eb8a8ae0c).

File: server/athenian/api/openapi/openapi.yaml (hunk 21 above)
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:850]: done [here](https://github.com/athenianco/athenian-api/pull/334/commits/6ce2316202b0863ecbb1a3a98385b49771ffc446).

File: server/athenian/api/openapi/openapi.yaml [hunk 25]
```diff
@@ -838,6 +838,219 @@ paths:
...
+        schema:
+          type: integer
+        style: simple
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Team'
+        x-body-name: body
+      responses:
```
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:984]: done [here](https://github.com/athenianco/athenian-api/pull/334/commits/57e0f929faa220ecdf10e91a922d70325f407f27).

File: server/athenian/api/openapi/openapi.yaml [hunk 26]
```diff
@@ -838,6 +838,219 @@ paths:
...
+        200:
+          content:
+            application/json:
+              schema:
+                type: object
+          description: Empty object indicates a successful operation.
```
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:990]: done [here](https://github.com/athenianco/athenian-api/pull/334/commits/eebbedc6dcb31817057c34cddedde07b6cd6e1a8).

File: server/athenian/api/openapi/openapi.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (dpordomingo) [reply server/athenian/api/openapi/openapi.yaml:886]: I don't need an answer. I'm just providing info :D

File: server/athenian/api/models/web/user.py [hunk 27]
```diff
@@ -131,7 +136,32 @@ def id(self, id: str):
...
+
+        :param login: The login of this User.
+        """
+        if login is None:
+            raise ValueError("Invalid value for `login`, must not be `None`")
+
+        self._login = login
+
+    @property
+    def native_id(self):
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/user.py:164]:
```suggestion
    def native_id(self) -> str:
```

**Comment by se7entyse7en** [comment]
Once approved I'll squash commits to have a sane history

File: server/athenian/api/openapi/openapi.yaml [hunk 28]
```diff
@@ -1083,6 +1308,152 @@ components:
...
+            updated: 2020-05-01T17:15:51Z
+          - login: github.com/se7entyse7en
+            id: github|5599208
+            native_id: 5599208
+            name: Lou Marvin Caraig
+            email: waren@athenian.co
+            picture: https://avatars1.githubusercontent.com/u/5599208?v=4
+            updated: 2020-05-05T18:19:05Z
+      properties:
+        id:
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1412]: Should we move `id` inside the `Team`? All the places that I see have them together.

I am also worried about how the team listing will work. If we accept a list of github logins then we will not always have an `id` - that's an Auth0 user ID. We cannot make it optional for the `User` either. It looks like we have to return back our `TeamMember`.

File: server/athenian/api/controllers/team_controller.py [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+async def create_team(request: AthenianWebRequest, body: dict) -> web.Response:
+    """Create a team.
+
+    :param request: HTTP request.
+    :param body: List of repositories to group.
+    """
+    pass
+
+
+async def delete_team(request: AthenianWebRequest, body: dict) -> web.Response:
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/team_controller.py:15]: Not all the handlers have a `body` - DELETE, for example.

Comment (vmarkovtsev) [inline server/athenian/api/controllers/team_controller.py:15]:
```suggestion
async def delete_team(request: AthenianWebRequest, id: int) -> web.Response:
```

File: server/athenian/api/openapi/openapi.yaml [hunk 30]
```diff
@@ -838,6 +838,231 @@ paths:
...
+        name: id
+        required: true
+        schema:
+          type: integer
+        style: simple
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/TeamUpdateRequest'
```
Comment (dpordomingo) [inline server/athenian/api/openapi/openapi.yaml:982]: Is it a good idea to be able to change the `team.id` within the `TeamUpdateRequest`? Since the team to be updated is already identified by the called URL, the `team.id` could be omitted.

File: server/athenian/api/openapi/openapi.yaml (hunk 28 above)
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:1412]: > Should we move id inside the Team? All the places that I see have them together.

Good point, indeed it was needed when the put needed to handle empty id. 👍 

> I am also worried about how the team listing will work. If we accept a list of github logins then we will not always have an id - that's an Auth0 user ID. We cannot make it optional for the User either. It looks like we have to return back our TeamMember.

Okay, I got the point. But this is related mostly to creation and update itself, isn't it? Instead of accepting a list of GH handles they should accept a list of ids.

**Comment by se7entyse7en** [comment]
@vmarkovtsev in addition the schema I also need to update all the places dealing with `User` to include the `login` attribute as well.

File: server/athenian/api/openapi/openapi.yaml (hunk 30 above)
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:982]: good point, since it's already in the path itself 👍

Comment (vmarkovtsev) [reply server/athenian/api/openapi/openapi.yaml:982]: This is a good point :+1:

File: server/athenian/api/openapi/openapi.yaml (hunk 28 above)
Comment (vmarkovtsev) [reply server/athenian/api/openapi/openapi.yaml:1412]: IDK what's the product decision here: allow any valid github login, allow any github login from the account's orgs or allow registered account users...

Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:1412]: I'd go strict, that is users that are member of the account, that would simplify a lot. Also in that case no need to go back to `TeamMember` since all members would be regular `User`s. \cc @warenlg

**Comment by se7entyse7en** [comment]
@vmarkovtsev @dpordomingo I updated the spec [here](https://github.com/athenianco/athenian-api/pull/334/commits/20fbb892bc072bc6865d9a08a6f82247686e634b):
- since as for Product requirement we need team  members that aren't users, a team now is not a list of `User`s, but a list of `TeamMember`s,
- for now, I reverted adding `login` to `User` since now it's not needed,
- added a `/contributors` endpoint to return all the contributors of an account.

@dpordomingo I didn't force-pushed, but it will be much easier to see the whole diff.

**Comment by se7entyse7en** [comment]
@dpordomingo I just realized that from `/contributors` I need to return also avatars, name, etc. I'm on it.

**Comment by se7entyse7en** [comment]
done [here](https://github.com/athenianco/athenian-api/pull/334/commits/b835cd1b88bef203239c907dec59f2a8488ff9ce).

**Comment by se7entyse7en** [comment]
Anyway @dpordomingo I'll know better what kind of errors (non-200) will be returned once I'll implement the endpoints.

File: server/athenian/api/openapi/openapi.yaml [hunk 31]
```diff
@@ -1082,6 +1082,12 @@ paths:
               schema:
                 $ref: '#/components/schemas/Contributors'
           description: List of all contributors belonging to the specified account.
+        403:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The user is forbidden to list contributors belonging to the specified account.
```
Comment (dpordomingo) [inline server/athenian/api/openapi/openapi.yaml:1090]: I guess than when implementing this, all account members will be allowed to list account team members (since regular users are already [able to change release settings](https://athenianco.atlassian.net/browse/ENG-696?focusedCommentId=12352), I'd say that they should be also able to list team members) 
wdyt @warenlg ?

File: server/athenian/api/controllers/contributors_controller.py [hunk 32]
```diff
@@ -0,0 +1,10 @@
+from aiohttp import web
+
+
+async def get_contributors(request: web.Request, id: int) -> web.Response:
+    """List all the contributors belonging to the specified account.
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/contributors_controller.py:5]: We should really-really post this in a separate PR. This is becoming an unmanageable monster.

File: server/athenian/api/models/web/contributor.py [hunk 33]
```diff
@@ -0,0 +1,130 @@
+from typing import Optional
+
+from athenian.api.models.web.base_model_ import Model
+
+
+class Contributor(Model):
+    """Contributor details."""
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/contributor.py:7]: If this is a spec description then we should be a bit more verbose here.
```suggestion
    """Details about a developer who contributed to some repositories owned by the account."""
```

File: server/athenian/api/openapi/openapi.yaml [hunk 34]
```diff
@@ -838,6 +838,278 @@ paths:
...
+        404:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The user does not belong to the specified account.
```
Comment (dpordomingo) [inline server/athenian/api/openapi/openapi.yaml:1096]: (not blocker) I'd say it should be returned a `403` if the user is not allowed to list the account contributors.

File: server/athenian/api/models/web/team_create_request.py [hunk 35]
```diff
@@ -0,0 +1,102 @@
...
+
+from athenian.api.models.web.base_model_ import Model
+
+
+class TeamCreateRequest(Model):
+    """Team creation request."""
+
+    def __init__(
+        self,
+            account: Optional[int] = None,
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team_create_request.py:11]: Indentation fault?

File: server/athenian/api/models/web/team_update_request.py [hunk 36]
```diff
@@ -0,0 +1,98 @@
+from typing import List, Optional
+
+from athenian.api.models.web.base_model_ import Model
+
+
+class TeamUpdateRequest(Model):
+    """Team update request."""
+
+    def __init__(
+            self,
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/team_update_request.py:10]: I think we should de-indent -4 spaces here

File: server/athenian/api/openapi/openapi.yaml [hunk 37]
```diff
@@ -838,6 +838,278 @@ paths:
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Contributors'
+          description: List of all contributors belonging to the specified account.
+        403:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The user is forbidden to list contributors belonging to the specified account.
```
Comment (vmarkovtsev) [inline server/athenian/api/openapi/openapi.yaml:1090]: Hence this should go in a separate PR

File: server/athenian/api/controllers/contributors_controller.py (hunk 32 above)
Comment (se7entyse7en) [reply server/athenian/api/controllers/contributors_controller.py:5]: Do you mind if we keep just the spec and the stub? The idea is that with this version of the spec everything needed for the settings page is ready. Then after I finished the implementation I'll move to change the filtering.

File: server/athenian/api/models/web/team_create_request.py (hunk 35 above)
Comment (se7entyse7en) [reply server/athenian/api/models/web/team_create_request.py:11]: yup 🤦

File: server/athenian/api/models/web/contributor.py (hunk 33 above)
Comment (se7entyse7en) [reply server/athenian/api/models/web/contributor.py:7]: added both here and in spec

**Comment by se7entyse7en** [comment]
@vmarkovtsev for better review I'd rather open separate PRs for each endpoint after this.

File: server/athenian/api/controllers/contributors_controller.py (hunk 32 above)
Comment (se7entyse7en) [reply server/athenian/api/controllers/contributors_controller.py:5]: This is actually also the first endpoint that I implemented that just needs to be opened 😄

File: server/athenian/api/openapi/openapi.yaml [hunk 38]
```diff
@@ -1371,7 +1371,7 @@ components:
       type: array
     Contributor:
       additionalProperties: false
-      description: Contributor details.
+      description: Details about a developer who contributed to some repositories owned by the account.
```
Comment (dpordomingo) [inline server/athenian/api/openapi/openapi.yaml:1374]: Can't be added new hires to their team till they make a contribution? Why not listing also organization members?

Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:1374]: It will behave the same way as `/filter/contributors` just without the time constraint. New hires and org members won't be shown in the filtering as well as now. Is this limitation ok @warenlg ?

File: server/athenian/api/openapi/openapi.yaml (hunk 34 above)
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:1096]: Since errors are minor things to be completely defined now, I'd keep them as they're now and I'll review all of them on each endpoint that I'll implement to make the errors defined in the spec reflect the implementation.

File: server/athenian/api/openapi/openapi.yaml (hunk 31 above)
Comment (se7entyse7en) [reply server/athenian/api/openapi/openapi.yaml:1090]: yup as described in ENG-726 all users can manage teams.

File: server/athenian/api/openapi/openapi.yaml (hunk 38 above)
Comment (warenlg) [reply server/athenian/api/openapi/openapi.yaml:1374]: Yes, it is ok, I don't want us to spend eng effort on new hires that have not contributed anywhere yet. The use case is too minor, and there is no need to handle it from our users.

File: server/athenian/api/models/web/user.py [hunk 39]
```diff
@@ -15,6 +15,7 @@ class User(Model):
     def __init__(
         self,
         id: Optional[str] = None,
+        login: Optional[str] = None,
```
Comment (vmarkovtsev) [inline server/athenian/api/models/web/user.py:18]: This must be covered in the tests + handled in `auth.py`. We can revert back if it is too much side work.

Comment (se7entyse7en) [reply server/athenian/api/models/web/user.py:18]: Ah indeed! Sorry, as I said [here](https://github.com/athenianco/athenian-api/pull/334#issuecomment-627276698) I actually reverted it back from the spec, but then forgot to revert the model itself. I'll do when cleaning up the history since it's not required now and would require extra work.

**Comment by se7entyse7en** [comment]
I cleaned up the history, I'll merge this once CI gives the green light.

File: server/athenian/api/models/web/user.py (hunk 39 above)
Comment (vmarkovtsev) [reply server/athenian/api/models/web/user.py]: :+1:

**Comment by dpordomingo** [comment]
@marvin Are these new endpoints working at staging?
```
curl -X GET "https://api-staging.athenian.co/v1/teams/1" -H "accept: application/json"
```
No response, no headers...
Imo it should return an empty array if no teams were found.

Then I tried to create a team:
```
curl 'https://api-staging.athenian.co/v1/team/create' \
    --data-binary '{
        "account":1,
        "name":"Engineers",
        "members":["github.com/dpordomingo","github.com/se7entyse7en"]
    }'
```
No response, no headers...
It should return the ID of the created group.

Then I tried to list the teams again to receive the created one
```
curl -X GET "https://api-staging.athenian.co/v1/teams/1" -H "accept: application/json"
```
No response, no headers...

**Comment by vmarkovtsev** [comment]
@dpordomingo No, they are not implemented yet.

**Comment by se7entyse7en** [comment]
Indeed, this PR was meant to just prepare the spec and add the stubs as minimal as possible to just make tests pass and be able to merge the spec. The first endpoint that will be ready is `/contributors/{id}`: https://github.com/athenianco/athenian-api/pull/344
